libs/lb/lb-rs/src/text/buffer.rs

```diff
@@ -0,0 +1,583 @@
+        let ops_a = diff(&self.external.text, &external_text_a);
+        let ops_b = diff(&self.external.text, &external_text_b);
+
+        let timestamp = Instant::now();
+        let base = self.external.seq;
+        self.ops
+            .meta
+            .extend(ops_a.iter().map(|_| OpMeta { timestamp, base }));
+        self.ops
+            .meta
+            .extend(ops_b.iter().map(|_| OpMeta { timestamp, base }));
+
+        self.ops
+            .all
+            .extend(ops_a.into_iter().map(Operation::Replace));
+        self.ops
+            .all
+            .extend(ops_b.into_iter().map(Operation::Replace));
+
+        self.update();
+        self.current.text
+    }
+
+    /// Applies all operations in the buffer's input queue
+    pub fn update(&mut self) -> Response {
+        // clear redo stack
+        //             v base        v current    v processed
+        // ops before: |<- applied ->|<- undone ->|<- queued ->|
+        // ops after:  |<- applied ->|<- queued ->|
+        let queue_len = self.base.seq + self.ops.len() - self.ops.processed_seq;
+        if queue_len > 0 {
+            let drain_range = self.current.seq..self.ops.processed_seq;
+            self.ops.all.drain(drain_range.clone());
+            self.ops.meta.drain(drain_range.clone());
+            self.ops.transformed.drain(drain_range.clone());
+            self.ops.transformed_inverted.drain(drain_range.clone());
+            self.ops.processed_seq = self.current.seq;
+        } else {
+            return Response::default();
+        }
+
+        // transform & apply
+        let mut result = Response::default();
+        for idx in self.current_idx()..self.current_idx() + queue_len {
+            let mut op = self.ops.all[idx].clone();
+            let meta = &self.ops.meta[idx];
+            self.transform(&mut op, meta);
+            self.ops.transformed_inverted.push(self.current.invert(&op));
+            self.ops.transformed.push(op.clone());
+            self.ops.processed_seq += 1;
+
+            result |= self.redo();
+        }
+
+        result
+    }
+
+    fn transform(&self, op: &mut Operation, meta: &OpMeta) {
+        let base_idx = meta.base - self.base.seq;
+        for transforming_idx in base_idx..self.ops.processed_seq {
+            let preceding_op = &self.ops.all[transforming_idx];
+            if let Operation::Replace(Replace {
+                range: preceding_replaced_range,
+                text: preceding_replacement_text,
+            }) = preceding_op
+            {
+                match op {
+                    Operation::Replace(Replace { range: transformed_range, .. })
+                    | Operation::Select(transformed_range) => {
+                        adjust_subsequent_range(
+                            *preceding_replaced_range,
+                            preceding_replacement_text.graphemes(true).count().into(),
+                            true,
```

Comment:
`true` here to advance the cursor when the replacement inserts text at the location of a subsequent cursor placement (new behavior)